src/main/kotlin/com/cosmotech/connector/adt/utils/AzureDigitalTwinsUtil.kt

```diff
@@ -168,16 +168,21 @@ class AzureDigitalTwinsUtil {
          * @return the modelInformationList completed
          */
         @JvmStatic
-        fun retrievePropertiesFromExtendedModel(modelInformationList: MutableList<DTDLModelInformation>):MutableList<DTDLModelInformation> {
+        fun retrievePropertiesFromBaseModels(modelInformationList: MutableList<DTDLModelInformation>):MutableList<DTDLModelInformation> {
             modelInformationList
-                .sortedBy { it.id }
-                .filter { it.isExtension }
-                .forEach { information ->
-                    val extendedModel =
-                        modelInformationList.find { information.baseModels?.contains(it.id) ?: false}
-                    extendedModel?.properties?.forEach { (key, value) ->
-                        information.properties.putIfAbsent(key, value)
-                    }
+                .filter { it.baseModels == null }
+                .forEach { baseModel ->
+                  fun fillExtendingModelsProperties(currentModel: DTDLModelInformation) {
```

Comment:
Local functions in a loop can incur some performance penalty. So I'd suggest declaring this recursive function outside of the loop in a `private` or `internal` function.